### PR TITLE
[C++]: C++17 test PR

### PR DESCRIPTION
**Pull request by pitrou**
Exercise some C++17 features to check platform support

### CODE REVIEW

**Comment by pitrou** [comment]
Small regression: our `string_view` backport features C++20 features such as `starts_with` and `ends_with`.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
Started a crossbow build at https://github.com/ursacomputing/crossbow/branches/all?query=build-327

**Comment by pitrou** [comment]
New crossbow build at https://github.com/ursacomputing/crossbow/branches/all?query=build-329

**Comment by pitrou** [comment]
I really like that `if constexpr` can help avoid clumsy template specialization with SFINAE (see example in `codegen_internal.h`).

**Comment by pitrou** [comment]
Problem: R 4.0.5 on CentOS7 does not define configuration variables for C++ versions higher than C++11:
```
# R CMD config CXX11 
g++
# R CMD config CXX14

# R CMD config CXX17

# R CMD config CXX11FLAGS
-g -O2
# R CMD config CXX14FLAGS

# R CMD config CXX17FLAGS

# R CMD config CXX11STD
-std=gnu++11
# R CMD config CXX14STD

# R CMD config CXX17STD

```

File: cpp/src/arrow/util/cache_internal.h [hunk 1]
```diff
@@ -174,22 +172,17 @@ template <template <typename...> class Cache, template <typename...> class Memoi
...
 }
 
 }  // namespace detail
 
 // Apply a LRU memoization cache to a callable.
 template <typename Func>
-static auto MemoizeLru(Func&& func, int32_t cache_capacity)
-    -> decltype(detail::Memoize<LruCache, detail::ThreadSafeMemoizer>(
-        std::forward<Func>(func), cache_capacity)) {
+static auto MemoizeLru(Func&& func, int32_t cache_capacity) -> auto {
```
Comment (bkietz) [inline cpp/src/arrow/util/cache_internal.h:185]:
```suggestion
static auto MemoizeLru(Func&& func, int32_t cache_capacity) {
```

File: cpp/src/arrow/util/cache_internal.h [hunk 2]
```diff
@@ -199,9 +192,7 @@ static auto MemoizeLru(Func&& func, int32_t cache_capacity)
 // A recommended usage is to declare per-thread caches using `thread_local`
 // (see cache_benchmark.cc).
 template <typename Func>
-static auto MemoizeLruThreadUnsafe(Func&& func, int32_t cache_capacity)
-    -> decltype(detail::Memoize<LruCache, detail::ThreadUnsafeMemoizer>(
-        std::forward<Func>(func), cache_capacity)) {
+static auto MemoizeLruThreadUnsafe(Func&& func, int32_t cache_capacity) -> auto {
```
Comment (bkietz) [inline cpp/src/arrow/util/cache_internal.h:195]:
```suggestion
static auto MemoizeLruThreadUnsafe(Func&& func, int32_t cache_capacity) {
```

**Comment by nealrichardson** [comment]
> Problem: R 4.0.5 on CentOS7 does not define configuration variables for C++ versions higher than C++11:

Bad config in the docker image? Or this is because it is gcc 4.8? (We have a crossbow job that does centos7 with devtoolset-8)

**Comment by pitrou** [comment]
What do you mean? These are the results of the `R` command.

**Comment by pitrou** [comment]
Ah, indeed I see this in the build logs:
```
-- Building using CMake version: 3.19.2
-- The C compiler identification is GNU 4.8.5
```
https://github.com/apache/arrow/pull/10414/checks?check_run_id=2687678135#step:9:69

**Comment by nealrichardson** [comment]
> What do you mean? These are the results of the `R` command.

Right, those commands read from R's Makeconf. If the compiler doesn't support C++14 or C++17, it's reasonable to expect that those makevars would be undefined.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 3]
```diff
@@ -67,44 +67,26 @@ constexpr Unsigned to_unsigned(T signed_) {
...
-struct AbsoluteValueChecked {
-  template <typename T, typename Arg>
-  static enable_if_signed_integer<T> Call(KernelContext*, Arg arg, Status* st) {
-    static_assert(std::is_same<T, Arg>::value, "");
-    if (arg == std::numeric_limits<Arg>::min()) {
-      *st = Status::Invalid("overflow");
+  static constexpr T Call(KernelContext*, T arg, Status* st) {
+    if constexpr (std::is_floating_point<T>::value) {
+      return std::fabs(arg);
+    } else if constexpr (std::is_signed<T>::value) {
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:76]: cpplint is failing for some of these: https://github.com/apache/arrow/pull/10414/checks?check_run_id=2687678079#step:7:2040

It looks like this has been fixed upstream: https://github.com/cpplint/cpplint/pull/136

File: cpp/src/arrow/testing/gtest_util.h [hunk 4]
```diff
@@ -24,6 +24,7 @@
 #include <functional>
 #include <memory>
 #include <string>
+#include <string_view>
```
Comment (bkietz) [inline cpp/src/arrow/testing/gtest_util.h:27]:
```suggestion
#include <string_view>
#include <optional>
```

Comment (bkietz) [reply cpp/src/arrow/testing/gtest_util.h:27]: We'll want to delete the `nonstd` string_view and optional impls to be sure we don't have both lying around. Similarly, we can remove `util::Variant` in favor of `std::variant`

**Comment by bkietz** [comment]
Debian 9 ([builds JNI](https://github.com/apache/arrow/pull/10414/checks?check_run_id=2687677911#step:8:2269)) has only gcc 6.3, which doesn't have `<string_view>`. https://wandbox.org/permlink/bijOi9BHbk3HXqiA In general it seems to lack a number of c++17 features

**Comment by kszucs** [comment]
> Debian 9 ([builds JNI](https://github.com/apache/arrow/pull/10414/checks?check_run_id=2687677911#step:8:2269)) has only gcc 6.3, which doesn't have `<string_view>`. https://wandbox.org/permlink/bijOi9BHbk3HXqiA In general it seems to lack a number of c++17 features

We should bump the debian version to 10 in those docker images (which is the default in `.env` but hardcoded in the java images). We may need to bump the `JDK` and `mvn` versions as well due to newer debian. If we wan't to keep builds for the current `JDK` and `mvn` versions then we can still parametrize that we test `debian-java` with `DEBIAN=9` and `debian-java-jni` with `DEBIAN=10`. 

I'm going to update the configuration.

**Comment by bkietz** [comment]
I've pushed two commits removing the util::string_view and util::optional aliases; some of the older compilers seemed to have trouble with those.

**Comment by kszucs** [comment]
@bkietz we have a couple more compile errors in the c-glib builds and on appveyor.

File: r/configure [hunk 5]
```diff
@@ -196,20 +196,10 @@ if [ "$ARROW_R_CXXFLAGS" ]; then
...
-  # now let's set the other two variables
-  CXX11FLAGS=`"${R_HOME}"/bin/R CMD config CXX11FLAGS`
-  CXX11STD=`"${R_HOME}"/bin/R CMD config CXX11STD`
-else
-  # For compatibility with R < 3.4, when these were called CXX1X
-  CXXCPP="`${R_HOME}/bin/R CMD config CXX1X` -E"
-  CXX11FLAGS=`"${R_HOME}"/bin/R CMD config CXX1XFLAGS`
-  CXX11STD=`"${R_HOME}"/bin/R CMD config CXX1XSTD`
-fi
+CXX17FLAGS=`"${R_HOME}"/bin/R CMD config CXX17FLAGS`
```
Comment (nealrichardson) [inline r/configure:199]: I don't think you actually want/need to delete this `if` block, but regardless the deletion was overzealous and we still need to define `CXXCPP` for the test command on L202. This should fix the Linux R builds (or the reason they're currently failing).

```suggestion
CXXCPP="`${R_HOME}/bin/R CMD config CXX17` -E"
CXX17FLAGS=`"${R_HOME}"/bin/R CMD config CXX17FLAGS`
```

Comment (bkietz) [reply r/configure:200]: Thanks for spotting this!

I'm not sure what the r < 3.4 block would be. Would CXX1X be able to compile c++17?

Comment (nealrichardson) [reply r/configure:200]: Possibly! We have a crossbow job that will tell us whether it works (if you restore the if block)

Comment (bkietz) [reply r/configure:200]: will do

**Comment by bkietz** [comment]
Thanks @jonkeane for your help figuring out Makevars

**Comment by bkietz** [comment]
@github-actions crossbow submit

**Comment by bkietz** [comment]
@github-actions crossbow submit test-*

**Comment by bkietz** [comment]
@github-actions crossbow submit -g r

**Comment by pitrou** [comment]
Submitted crossbow builds:
https://github.com/ursacomputing/crossbow/branches/all?query=build-360
https://github.com/ursacomputing/crossbow/branches/all?query=build-361

**Comment by jeroen** [comment]
CRAN targets MacOS 10.13 (xcode10.1). Did anyone test if this version of xcode supports C++17?

**Comment by nealrichardson** [comment]
@jeroen there are a number of packages on CRAN that use C++17 already, aren't there?

**Comment by nealrichardson** [comment]
hmm, fewer than I thought:

```
> table(grepl("C++17", tools::CRAN_package_db()$SystemRequirements, fixed = TRUE))

FALSE  TRUE 
17691     3 
```

**Comment by jeroen** [comment]
Yes I think so, but that does not guarantee that the given compilers support the full c++17 feature set. Perhaps it would be good to add a travis build with:

```
os: osx
osx_image: xcode10.1
```

Afaik there is no other CI supporting older versions of MacOS.

**Comment by jeroen** [comment]
I tried building this PR on macos 10.13 but it failed with:

```
[ 16%] Building CXX object src/arrow/dataset/CMakeFiles/arrow_dataset_objlib.dir/file_ipc.cc.o
cd /tmp/apache-arrow-20210608-10068-17s4hhj/build/src/arrow/dataset && /Users/builder/Desktop/brew/Library/Homebrew/shims/mac/super/clang++ -DARROW_DS_EXPORTING -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_JEMALLOC -DARROW_JEMALLOC_INCLUDE_DIR="" -DARROW_MIMALLOC -DARROW_WITH_LZ4 -DARROW_WITH_RE2 -DARROW_WITH_SNAPPY -DARROW_WITH_TIMING_TESTS -DARROW_WITH_UTF8PROC -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DURI_STATIC_BUILD -DUTF8PROC_STATIC -I/tmp/apache-arrow-20210608-10068-17s4hhj/build/src -I/tmp/apache-arrow-20210608-10068-17s4hhj/cpp/src -I/tmp/apache-arrow-20210608-10068-17s4hhj/cpp/src/generated -I/tmp/apache-arrow-20210608-10068-17s4hhj/cpp/src/parquet -isystem /tmp/apache-arrow-20210608-10068-17s4hhj/cpp/thirdparty/flatbuffers/include -isystem /Users/builder/Desktop/brew/Cellar/thrift/0.11.0/include -isystem /tmp/apache-arrow-20210608-10068-17s4hhj/build/jemalloc_ep-prefix/src -isystem /tmp/apache-arrow-20210608-10068-17s4hhj/build/mimalloc_ep/src/mimalloc_ep/lib/mimalloc-1.6/include -isystem /tmp/apache-arrow-20210608-10068-17s4hhj/build/rapidjson_ep/src/rapidjson_ep-install/include -isystem /tmp/apache-arrow-20210608-10068-17s4hhj/build/xsimd_ep/src/xsimd_ep-install/include -isystem /Users/builder/Desktop/brew/Cellar/lz4/1.8.3/include -isystem /Users/builder/Desktop/brew/Cellar/zstd/1.3.5/include -isystem /tmp/apache-arrow-20210608-10068-17s4hhj/build/re2_ep-install/include -isystem /tmp/apache-arrow-20210608-10068-17s4hhj/build/utf8proc_ep-install/include -isystem /tmp/apache-arrow-20210608-10068-17s4hhj/cpp/thirdparty/hadoop/include -Qunused-arguments -fcolor-diagnostics -O3 -DNDEBUG  -Wall -Wno-unknown-warning-option -Wno-pass-failed -stdlib=libc++ -msse4.2  -DNDEBUG -fPIC -std=c++17 -o CMakeFiles/arrow_dataset_objlib.dir/file_ipc.cc.o -c /tmp/apache-arrow-20210608-10068-17s4hhj/cpp/src/arrow/dataset/file_ipc.cc
/tmp/apache-arrow-20210608-10068-17s4hhj/cpp/src/arrow/dataset/file_base.cc:77:38: error: call to unavailable member function 'value': introduced in macOS 10.14
    actual_compression = compression.value();
                         ~~~~~~~~~~~~^~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:942:27: note: candidate function has been explicitly made unavailable
    constexpr value_type& value() &
                          ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:933:33: note: candidate function has been explicitly made unavailable
    constexpr value_type const& value() const&
                                ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:951:28: note: candidate function not viable: no known conversion from 'optional<...>' to 'optional<...>' for object argument
    constexpr value_type&& value() &&
                           ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/optional:960:34: note: candidate function not viable: no known conversion from 'optional<...>' to 'const optional<...>' for object argument
    constexpr value_type const&& value() const&&
                                 ^
```

**Comment by jeroen** [comment]
How exactly do I do that? I tried literally substituting that line and that errors with:

```
file_base.cc:77:26: error: 'compression' is not a class, namespace, or enumeration
    actual_compression = compression::optional::operator*;
```

**Comment by nealrichardson** [comment]
Closing as stale

**Pull request closed without merging**